read-fonts/src/tables/aat.rs

```diff
@@ -0,0 +1,619 @@
+            }
+            Lookup::Format8(lookup) => index
+                .checked_sub(lookup.first_glyph())
+                .and_then(|ix| {
+                    lookup
+                        .value_array()
+                        .get(ix as usize)
+                        .map(|val| T::from_u16(val.get()))
+                })
+                .ok_or(ReadError::OutOfBounds),
+            Lookup::Format10(lookup) => {
+                let ix = index
+                    .checked_sub(lookup.first_glyph())
+                    .ok_or(ReadError::OutOfBounds)? as usize;
+                let unit_size = lookup.unit_size() as usize;
+                let offset = ix * unit_size;
+                let mut cursor = FontData::new(lookup.values_data()).cursor();
+                cursor.advance_by(offset);
+                let val = match unit_size {
+                    1 => cursor.read::<u8>()? as u32,
+                    2 => cursor.read::<u16>()? as u32,
+                    4 => cursor.read::<u32>()?,
+                    _ => {
+                        return Err(ReadError::MalformedData(
+                            "invalid unit_size in format 10 AAT lookup table",
+                        ))
+                    }
+                };
+                Ok(T::from_u32(val))
+            }
+        }
+    }
+}
+
+impl<'a, T> FontRead<'a> for TypedLookup<'a, T> {
+    fn read(data: FontData<'a>) -> Result<Self, ReadError> {
+        Ok(Self {
+            lookup: Lookup::read(data)?,
+            _marker: std::marker::PhantomData,
+        })
+    }
+}
+
+#[cfg(feature = "traversal")]
+impl<'a, T> SomeTable<'a> for TypedLookup<'a, T> {
+    fn type_name(&self) -> &str {
+        "TypedLookup"
+    }
+
+    fn get_field(&self, idx: usize) -> Option<Field<'a>> {
+        self.lookup.get_field(idx)
+    }
+}
+
+/// Trait for values that can be read from lookup tables.
+pub trait LookupValue: Copy + Scalar + bytemuck::AnyBitPattern {
+    fn from_u16(v: u16) -> Self;
+    fn from_u32(v: u32) -> Self;
+}
+
+impl LookupValue for u16 {
+    fn from_u16(v: u16) -> Self {
+        v
+    }
+
+    fn from_u32(v: u32) -> Self {
+        // intentionally truncates
+        v as _
+    }
+}
+
+impl LookupValue for u32 {
+    fn from_u16(v: u16) -> Self {
+        v as _
+    }
+
+    fn from_u32(v: u32) -> Self {
+        v
+    }
+}
+
+impl LookupValue for GlyphId {
+    fn from_u16(v: u16) -> Self {
+        GlyphId::from(v)
+    }
+
+    fn from_u32(v: u32) -> Self {
+        // intentionally truncates
+        GlyphId::from(v as u16)
+    }
+}
+
+pub type LookupU16<'a> = TypedLookup<'a, u16>;
+pub type LookupU32<'a> = TypedLookup<'a, u32>;
+pub type LookupGlyphId<'a> = TypedLookup<'a, GlyphId>;
+
+/// Entry in an (extended) state table.
+pub struct StateEntry<T = ()> {
+    /// Index of the next state.
+    pub new_state: u16,
+    /// Flag values are table specific.
+    pub flags: u16,
+    /// Payload is table specific.
+    pub payload: T,
```

Comment:
I'd like to know more around what the allowed values are, what the invariants are, etc. e.g,

- for a given table, is `T` always identical? At what point do we know the value?
- what are the possible values for `T`, overall?


This would help me make a more informed assessment of whether or not this is the best way of expressing what we want.

Reply:
I think I answered this above but I can elaborate if not